cmd/semaphore/daemon/nice_errors.go

```diff
@@ -0,0 +1,216 @@
+package daemon
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/jexia/semaphore/pkg/references"
+)
+
+const (
+	UnresolvedCallCode         = "UnresolvedCall"
+	unresolvedParamsCode       = "UnresolvedParams"
+	unresolvedPropertyCode     = "UnresolvedProperty"
+	unresolvedParameterMapCode = "UnresolvedParameterMap"
+	unableToResolveFlowCode    = "UnableToResolveFlow"
+	undefinedReferenceCode     = "UndefinedReference"
+	undefinedResourceCode      = "UndefinedResource"
+	unresolvedOnErrorCode      = "UnresolvedOnError"
+	genericErrorCode           = "GenericError"
+	unresolvedNodeCode         = "UnresolvedNode"
+)
+
+type NiceError struct {
+	Code    string                 `json:"code"`
+	Message string                 `json:"message"`
+	Details map[string]interface{} `json:"details"`
+}
+
+// NewNiceError takes an internal error and returns a new one that is much nicer to users.
+func NewNiceError(err error) error {
+	stack := humanizeErrors(err)
+
+	return errors.New(formatNiceErrors(stack))
+}
+
+// format stack of Nice Errors with a nice format. For example, it might be a simple json.Marshal call,
+// but here we do some friendly magic.
+func formatNiceErrors(stack []NiceError) string {
+	o := bytes.NewBufferString("\n")
+
+	for _, nice := range stack {
+		fmt.Fprintf(o, "(%s) %s\n", nice.Code, nice.Message)
+		if nice.Details != nil {
+			for k, v := range nice.Details {
+				value, _ := json.Marshal(v)
+				fmt.Fprintf(o, "\t%s: %s\n", k, value)
+			}
+		}
+	}
+
+	return o.String()
+}
+
+func humanizeErrors(next error) []NiceError {
+	var (
+		stack []NiceError
+	)
+
+	for next != nil {
+		stack = append(stack, humanizeError(next))
+		next = errors.Unwrap(next)
+	}
+
+	return stack
+}
+
+// check the error and build a friendly error message
+func humanizeError(inner error) (nice NiceError) {
+	var (
+		errUnresolvedFlow         references.ErrUnresolvedFlow
+		errUndefinedReference     references.ErrUndefinedReference
+		errUndefinedResource      references.ErrUndefinedResource
+		errUnresolvedOnError      references.ErrUnresolvedOnError
+		errUnresolvedNode         references.ErrUnresolvedNode
+		errUnresolvedParameterMap references.ErrUnresolvedParameterMap
+		errUnresolvedProperty     references.ErrUnresolvedProperty
+		errUnresolvedParams       references.ErrUnresolvedParams
+		errUnresolvedCall         references.ErrUnresolvedCall
+	)
+
+	switch {
```

Comment:
Maybe it is better to keep the formatting of "human" errors inside the error itself to avoid having an enormous switch case? A generic interface including a single method returning a `NiceError` would be called allowing to keep the error messages as close to the error as possible.